protocol/protobuf/message.proto

```diff
@@ -0,0 +1,172 @@
+syntax = "proto3";
+
+package protobuf;
+
+message SyncContact {
+  // The public key of the contact to be synced
+  string public_key = 1;
+  // The profile image of the contact
+  string profile_image = 2;
+  // The ens_name of the contact
+  string ens_name = 3;
+  // The system tags for a given contact
+  repeated SystemTags system_tags = 4;
+
+  enum SystemTags {
+    ADDED_BY_US = 0;
+    BLOCKED = 1;
+    ADDED_BY_THEM = 2;
+  }
+}
+
+message SyncPublicChat {
+  // The chat_id of the chat to be synced
+  string chat_id = 1;
+}
+
+message SyncAccount {
+  // Our profile image
+  string profile_image = 1;
+  // The time we last updated the account
+  uint64 last_updated = 2;
+}
+
+message SyncInstallation {
+  // The contacts to be synced
+  repeated SyncContact contacts = 1;
+  // The public chats to be synced
+  repeated SyncPublicChat public_chats = 2;
+  // Our own account
+  repeated SyncAccount account = 3;
+}
+
+message PairInstallation {
+  // The user-set device name
+  string device_name = 1;
+  // The installation-id of the device
+  string installation_id = 2;
+  // The type of device
+  DeviceType device_type = 3;
+  enum DeviceType {
+    ANDROID = 0;
+    IOS = 1;
+    DESKTOP = 2;
+  }
+}
+
+message ContactRequest {
+  // The ens name of the contact
+  string ens_name = 1;
+  // The profile image of the contact
+  string profile_image = 2;
+}
+
+message GroupChatEvent {
+  // Lamport timestamp of the event
+  uint64 clock = 1;
+  // List of public keys of objects of the action
+  repeated string members = 2;
+  // Name of the chat for the CHAT_CREATED/NAME_CHANGED event types
+  string name = 3;
+  // The type of the event
+  EventType type = 4;
+
+  enum EventType {
+    CHAT_CREATED = 0;
+    NAME_CHANGED = 1;
+    MEMBERS_ADDED = 2;
+    MEMBER_JOINED = 3;
+    MEMBER_REMOVED = 4;
+    ADMINS_ADDED = 5;
+    ADMIN_REMOVED = 6;
+  }
+}
+
+message MembershipUpdateMessage {
+  // The chat id of the private group chat
+  string chat_id = 1;
+  // A list of updates for this group chat
+  repeated MembershipUpdate updates = 2;
+  // An optional chat message
+  ChatMessage message = 3;
+}
+
+message MembershipUpdate {
+  // Signature of the concatenated array bytes of the events
+  string signature = 1;
+  // The protobuf encoded events
+  repeated bytes events = 2;
```

Comment:
Why this one is not of `GroupChatEvent` type?

Reply:
I haven't settled on this, but basically it's because if we make it`bytes`, we can more easily change signature scheme without breaking compatibility (also it simplifies signing).

How it is now is that we concatenate each field and we sign the concatenated JSON fields.
i.e (simplified):

```
Sign(field_a + field_b + field_c)
```
and we send the signature.

The issue with this approach is that if we add `field_d`, it will break compatibility (unless we propagate 2 different signatures, one with and one without), as old client will not be using the new field for signature.

One way to overcome this, is to sign the encoded payload (same as we do for the whole message), so that any change inside the payload won't break the signature, at the expenses of having to keep the event payload for propagation.